File: apps/csharp/tests/Ch07/VoteCounterTests.cs
using GrokkingConcurrency.CSharp.Ch07;

namespace GrokkingConcurrency.CSharp.Tests.Ch07;

/// <summary>
/// Fork/Join パターンによる投票集計のテスト
/// </summary>
public class VoteCounterTests
{
    [Fact]
    public void ShouldCountVotes()
    {
        var votes = new List<int> { 1, 2, 1, 3, 2, 1, 1, 2, 3, 1 };

        var result = VoteCounter.CountVotes(votes);

        Assert.Equal(5, result[1]);
        Assert.Equal(3, result[2]);
        Assert.Equal(2, result[3]);
    }

    [Fact]
    public void ShouldHandleEmptyList()
    {
        var votes = new List<int>();

        var result = VoteCounter.CountVotes(votes);

        Assert.Empty(result);
    }

    [Fact]
    public void ShouldCountVotesInParallel()
    {
        // 10万票を生成
        var votes = Enumerable.Range(0, 100000)
            .Select(i => i % 10 + 1)
            .ToList();

        var result = VoteCounter.CountVotesParallel(votes, 4);

        // 各候補者に1万票ずつ
        for (var i = 1; i <= 10; i++)
        {
            Assert.Equal(10000, result[i]);
        }
    }

    [Fact]
    public void ShouldMergeResults()
    {
        var result1 = new Dictionary<int, int> { { 1, 5 }, { 2, 3 } };
        var result2 = new Dictionary<int, int> { { 1, 3 }, { 3, 4 } };

        var merged = VoteCounter.MergeResults(result1, result2);

        Assert.Equal(8, merged[1]);
        Assert.Equal(3, merged[2]);
        Assert.Equal(4, merged[3]);
    }
}

File: apps/fsharp/tests/VoteCounterTests.fs
module VoteCounterTests

open Xunit
open GrokkingConcurrency.Ch07.VoteCounter

[<Fact>]
let ``countVotes counts votes for each candidate`` () =
    let votes = ["A"; "B"; "A"; "C"; "A"; "B"]
    let result = countVotes votes
    Assert.Equal(3, Map.find "A" result)
    Assert.Equal(2, Map.find "B" result)
    Assert.Equal(1, Map.find "C" result)

[<Fact>]
let ``countVotesParallel counts votes using fork-join`` () =
    let votes = List.replicate 100 "A" @ List.replicate 50 "B" @ List.replicate 30 "C"
    let result = countVotesParallel votes
    Assert.Equal(100, Map.find "A" result)
    Assert.Equal(50, Map.find "B" result)
    Assert.Equal(30, Map.find "C" result)

[<Fact>]
let ``countVotesParallel handles empty list`` () =
    let result = countVotesParallel []
    Assert.True(Map.isEmpty result)

[<Fact>]
let ``mergeResults combines two vote counts`` () =
    let a = Map.ofList [("A", 10); ("B", 5)]
    let b = Map.ofList [("A", 5); ("C", 3)]
    let result = mergeResults a b
    Assert.Equal(15, Map.find "A" result)
    Assert.Equal(5, Map.find "B" result)
    Assert.Equal(3, Map.find "C" result)
